File: apple/CaCompteKit/Tests/CatalogTests/SeededGenerator.swift
/// SplitMix64 — générateur déterministe : un test qui échoue redevient reproductible à partir
/// de son seed, contrairement à `SystemRandomNumberGenerator`. Copie de
/// `DomainTests/SeededGenerator.swift` : `CatalogTests` ne dépend pas de `DomainTests` (voir
/// `Package.swift`), donc pas de partage direct entre cibles de test.
struct SeededGenerator: RandomNumberGenerator {
  private var state: UInt64

  init(seed: Int) {
    state = UInt64(bitPattern: Int64(seed)) &+ 0x9E37_79B9_7F4A_7C15
  }

  mutating func next() -> UInt64 {
    state &+= 0x9E37_79B9_7F4A_7C15
    var z = state
    z = (z ^ (z >> 30)) &* 0xBF58_476D_1CE4_E5B9
    z = (z ^ (z >> 27)) &* 0x94D0_49BB_1331_11EB
    return z ^ (z >> 31)
  }
}

File: android/catalog/src/test/kotlin/com/quimene/catalog/testing/SeededGenerator.kt
package com.quimene.catalog.testing

/**
 * Miroir de `SeededGenerator.swift` (`CatalogTests`) — SplitMix64, générateur déterministe : un
 * test qui échoue redevient reproductible à partir de son seed, contrairement à un générateur
 * système. Pour les petits seeds réellement utilisés par les tests (0..49), produit la même
 * séquence bit-à-bit que la version Swift (même élargissement signe-préservant vers 64 bits).
 */
class SeededGenerator(
    seed: Int,
) {
    private var state: ULong = seed.toLong().toULong() + 0x9E3779B97F4A7C15uL

    fun nextULong(): ULong {
        state += 0x9E3779B97F4A7C15uL
        var z = state
        z = (z xor (z shr 30)) * 0xBF58476D1CE4E5B9uL
        z = (z xor (z shr 27)) * 0x94D049BB133111EBuL
        return z xor (z shr 31)
    }

    fun nextInt(range: IntRange): Int {
        val span = (range.last - range.first + 1).toULong()
        return (nextULong() % span).toInt() + range.first
    }

    fun nextBoolean(): Boolean = nextULong() % 2uL == 0uL

    fun <T> randomElement(items: List<T>): T = items[nextInt(items.indices)]
}
